src/quantization/sq4_uniform_quantizer.h

```diff
@@ -0,0 +1,245 @@
+
+    inline float
+    ComputeImpl(const uint8_t* codes1, const uint8_t* codes2) const;
+
+    inline void
+    ProcessQueryImpl(const DataType* query, Computer<SQ4UniformQuantizer>& computer) const;
+
+    inline void
+    ComputeDistImpl(Computer<SQ4UniformQuantizer>& computer,
+                    const uint8_t* codes,
+                    float* dists) const;
+
+private:
+    DataType lower_bound_;
+    DataType diff_;
+    std::unordered_map<std::string, uint64_t> offsets_;
+};
+
+template <MetricType Metric>
+SQ4UniformQuantizer<Metric>::SQ4UniformQuantizer(int dim)
+    : Quantizer<SQ4UniformQuantizer<Metric>>(dim) {
+    lower_bound_ = std::numeric_limits<DataType>::max();
+    diff_ = std::numeric_limits<DataType>::min();
+
+    this->code_size_ = 0;
+
+    offsets_[OFFSET_KEY_CODE] = this->code_size_;
+    this->code_size_ += (dim + 1) >> 1 << 1;
+
+    if (Metric == MetricType::METRIC_TYPE_L2SQR or Metric == MetricType::METRIC_TYPE_COSINE) {
+        offsets_[OFFSET_KEY_NORM] = this->code_size_;
+        this->code_size_ += sizeof(norm_type);  // norm of vector
+    }
+
+    if (Metric == MetricType::METRIC_TYPE_IP or Metric == MetricType::METRIC_TYPE_COSINE) {
+        offsets_[OFFSET_KEY_SUM] = this->code_size_;
+        this->code_size_ += sizeof(sum_type);  // sum of vector
+    }
+}
+
+template <MetricType Metric>
+bool
+SQ4UniformQuantizer<Metric>::TrainImpl(const DataType* data, uint64_t count) {
+    if (data == nullptr) {
+        return false;
+    }
+
+    for (uint32_t i = 0; i < count; i++) {
+        for (uint32_t d = 0; d < this->dim_; d++) {
+            auto val = data[i * this->dim_ + d];
+            if (val > diff_) {
+                diff_ = val;
+            }
+            if (val < lower_bound_) {
+                lower_bound_ = val;
+            }
+        }
+    }
+
+    for (uint32_t d = 0; d < this->dim_; d++) {
+        diff_ -= lower_bound_;
+    }
+
+    this->is_trained_ = true;
+    return true;
+}
+
+template <MetricType Metric>
+bool
+SQ4UniformQuantizer<Metric>::EncodeOneImpl(const DataType* data, uint8_t* codes) const {
+    float delta;
+    uint8_t scaled;
+    norm_type norm = 0;
+    sum_type sum = 0;
+
+    for (uint32_t d = 0; d < this->dim_; d++) {
+        delta = ((data[d] - lower_bound_) / diff_);
+        if (delta < 0.0) {
+            delta = 0;
+        }
+        if (delta > 0.999) {
+            delta = 1;
+        }
+        scaled = 15 * delta;
+
+        if (d & 1) {
+            codes[offsets_.at(OFFSET_KEY_CODE) + d / 2] |= scaled << 4;
+        } else {
+            codes[offsets_.at(OFFSET_KEY_CODE) + d / 2] = 0;
+            codes[offsets_.at(OFFSET_KEY_CODE) + d / 2] |= scaled;
+        }
+        norm += scaled * scaled;
+        sum += data[d];
+    }
+
+    if (Metric == MetricType::METRIC_TYPE_L2SQR or Metric == MetricType::METRIC_TYPE_COSINE) {
+        *(norm_type*)(codes + offsets_.at(OFFSET_KEY_NORM)) = norm;
```

Comment:
offsets_.at(OFFSET_KEY_NORM) -> offsets_[OFFSET_KEY_NORM]
unify code style

Reply:
here use .at in const method to avoid any modification of offsets_

Reply:
vector has `operator[] const` override version

Reply:
remove unordered_map<> offsets_ and use three variables offset_code_, offset_norm_ and offset_sum_ to avoid hash computations in map.